storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1764 @@ buf_block_t *fsp_get_latched_xdes_page(
+    if (m_new_state == XDES_FSEG)
+      return alloc_from_fseg_complete();
+    return alloc_from_free_frag_complete();
+  }
+
+  /** Assign the fragment slot of the index node.
+  This step should be done after removing the old page
+  because there is a possiblity that FRAGMENT ARRAY
+  could be full. */
+  void assign_frag_slot()
+  {
+    if ((!m_need_segment && m_new_state == XDES_FREE) ||
+	m_new_state == XDES_FULL_FRAG ||
+        m_new_state == XDES_FREE_FRAG)
+      fseg_set_nth_frag_page_no(m_inode, m_iblock,
+                                fseg_find_free_frag_page_slot(m_inode),
+				m_new_page, m_mtr);
+  }
+
+  /** Restore the page modified during page allocation */
+  void restore_old_pages()
+  {
+    for (auto &it : m_old_pages)
+    {
+      buf_block_t *block=
+	m_mtr->get_already_latched(page_id_t{0, it.first},
+                                   MTR_MEMO_PAGE_SX_FIX);
+      ut_ad(block);
+      memcpy_aligned<UNIV_PAGE_SIZE_MIN>(
+        block->page.frame, it.second->page.frame, srv_page_size);
+    }
+  }
+};
+
+/** Steps for removing the page from old extent:
+
+(1) If the xdes_get_state(m_old_descr) is XDES_FSEG then
+   (1.1) If the number of used pages in descriptor is
+         FSP_EXTENT_SIZE then move the extent descriptor
+         from FSEG_FULL to FSEG_NOT_FULL
+   (1.2) If the number of used page in descriptor to 0 then
+         move the extent descriptor from FSEG_NOT_FULL to FSP_FREE
+   (1.3) Free the page and mark the XDES_FREE_BIT of the
+         respective page in current extent descriptor
+
+(2) If the xdes_get_stats(m_old_descr) is
+    XDES_FREE_FRAG / XDES_FULL_FRAG then
+
+    (2.1) If the number of used pages in extent descriptor
+          is FSP_EXTENT_SIZE then move the extent descriptor
+          from FSP_FULL_FRAG to FSP_FREE_FRAG list
+    (2.2) If the number of used pages in extent descriptor is 0 then
+          move the extent descriptor from FSP_FREE_FRAG to FSP_FREE
+    (2.3) Free the page and mark the XDES_FREE_BIT of the respective
+          page in current extent descriptor */
+class pageRemover final
+{
+private:
+  /* System tablespace */
+  fil_space_t *m_space=  fil_system.sys_space;
+  /* Header block for the tablespace */
+  buf_block_t *m_header_block= nullptr;
+  /* Index node block */
+  buf_block_t *m_iblock= nullptr;
+  /* Index node */
+  fseg_inode_t *m_inode= nullptr;
+  /* offset of index node within index node page*/
+  uint16_t m_ioffset= 0;
+  /* Maximum free limit of the tablespace */
+  uint32_t m_free_limit= 0;
+  /* Extent size */
+  uint32_t m_extent_size= 0;
+
+  /* Page to be removed */
+  uint32_t m_page_no= 0;
+  /* Old block extent descriptor page */
+  buf_block_t *m_old_xdes= nullptr;
+  /* Old block descriptor */
+  xdes_t *m_old_descr= nullptr;
+  /* Old block descriptor offset with descriptor page */
+  uint16_t m_xoffset= 0;
+  /* Old descriptor state */
+  uint32_t m_old_state= 0;
+  /* Mini-transaction for removal of pages */
+  mtr_t *m_mtr;
```

Comment:
This and any other data members that should be constant after the constructor has been invoked must be declared `const`:
```c++
  mtr_t *const m_mtr;
```